File: src/Java/algorithms/mathandprob/NextPermutation.java
package algorithms.mathandprob; 

import java.util.*; 

public class NextPermutation {
    /* Function: Find the next permutation of an array. If is the last perm, return the first perm. 
     * Solution: Find the first dis-order element from the end, then swap it with the next larger value, 
     *     and sort the elements after it. 
     */ 
    
    public void nextPermutation(int[] num) {
        if (num == null || num.length <= 1) {
            return; 
        }
        
        int i = num.length - 2; 
        while (i >= 0 && num[i] >= num[i + 1]) {
            --i; 
        }
        
        if (i == -1) {
            reverse(num, 0); 
            return; 
        }
        
        for (int j = num.length - 1; j > i; --j) {
            if (num[j] > num[i]) {
                swap(num, i, j); 
                break; 
            }
        }
        
        reverse(num, i + 1); 
    }
    
    private void reverse(int[] num, int st) { // Since they are sorted in decrease order. 
        int ed = num.length - 1; 
        while (st < ed) {
            swap(num, st++, ed--); 
        }
    }
    
    private void swap(int[] num, int a, int b) {
        int tmp = num[a]; 
        num[a] = num[b]; 
        num[b] = tmp; 
    }
}

File: src/CPP/src/leetcode/NextPermutation.cpp
/*
 * [Source] https://leetcode.com/problems/next-permutation/
 * [Difficulty]: Medium
 * [Tag]: Array
 */

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// [Solution]: Find the first number that is larger than its successor from end, swap it with the first number that is larger than it, and sort numbers after it.
// [Corner Case]: dup. last num
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        if (nums.empty())
            return;

        int p = nums.size() - 2;
        for (; p >= 0 && nums[p] >= nums[p + 1]; p--);

        if (p >= 0) {
            int i = p + 1;
            for (; i < (int)nums.size() && nums[p] < nums[i]; i++);
            swap(nums[i - 1], nums[p]);
        }
        reverse(nums.begin() + p + 1, nums.end());
    }
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/mathandprob/NextPermutation.java
 */

/* Java solution
public class Solution {
    public void nextPermutation(int[] num) {
        if(num == null || num.length <= 1) {
            return; 
        }
        int i, j; 
        for(i = num.length - 2; i >= 0; --i) {
            if(num[i] < num[i + 1]) {
                break; 
            }
        }
        if(i < 0) { // Alreay at the end. 
            Arrays.sort(num); 
            return; 
        }
        int key = num[i]; 
        for(j = num.length - 1; j >= 0; --j) {
            if(num[j] > key) {
                break; // At least i + 1 
            }
        }
        // swap
        num[i] = num[j]; 
        num[j] = key; 
        Arrays.sort(num, i + 1, num.length); 
    }
}
 */

int main() {
    Solution sol;

    //vector<int> vec = {4,5,3,2,1};
    //vector<int> vec = {3,2,1};
    //vector<int> vec = {1,1,2};
    //vector<int> vec = {1,2,1};
    //vector<int> vec = {1,2,2};
    vector<int> vec = {2,1,2};
    sol.nextPermutation(vec);
    for (int &num : vec)
        cout << num << ", ";
    cout << endl;

    return 0;
}
